File: backend/models/pricing.py
"""Pydantic models for pricing data."""

from datetime import datetime
from typing import Dict, List, Optional
from pydantic import BaseModel


class Pricing(BaseModel):
    """Price info (USD per million tokens/units)."""

    input: Optional[float] = None
    output: Optional[float] = None
    cached_input: Optional[float] = None
    cached_write: Optional[float] = None
    reasoning: Optional[float] = None
    image_input: Optional[float] = None
    audio_input: Optional[float] = None
    audio_output: Optional[float] = None
    embedding: Optional[float] = None


class BatchPricing(BaseModel):
    """Batch processing discounted prices."""

    input: Optional[float] = None
    output: Optional[float] = None


class ModelPricing(BaseModel):
    """Complete pricing info for a single model."""

    id: str  # Unique: "{provider}:{model_id}"
    provider: str  # aws_bedrock, openai, azure, etc.
    model_id: str  # Original model ID
    model_name: str  # Display name
    pricing: Pricing
    batch_pricing: Optional[BatchPricing] = None
    context_length: Optional[int] = None
    max_output_tokens: Optional[int] = None
    is_open_source: Optional[bool] = None  # True if weights are downloadable
    capabilities: List[str] = []  # ["text", "vision", "audio", "embedding"]
    input_modalities: List[str] = []  # ["text", "image", "audio", "video", "file"]
    output_modalities: List[str] = []  # ["text", "image", "audio", "video", "embedding"]
    last_updated: datetime


class PricingDatabase(BaseModel):
    """JSON file root structure."""

    version: str = "1.0"
    last_refresh: datetime
    models: List[ModelPricing] = []


class ProviderInfo(BaseModel):
    """Provider metadata for API response."""

    name: str
    display_name: str
    model_count: int
    last_updated: Optional[datetime] = None


class ProviderFile(BaseModel):
    """Structure for per-provider JSON file (providers/*.json)."""

    provider: str
    last_updated: datetime
    models: List[ModelPricing] = []


class ProviderIndexEntry(BaseModel):
    """Entry in the index file for a single provider."""

    file: str  # Relative path like "providers/openai.json"
    model_count: int
    last_updated: datetime


class IndexFile(BaseModel):
    """Structure for index.json - tracks all provider files."""

    version: str = "2.0"
    last_refresh: datetime
    providers: Dict[str, ProviderIndexEntry] = {}
    total_models: int = 0

File: frontend/src/types/pricing.ts
export interface Pricing {
  input: number | null;
  output: number | null;
  cached_input: number | null;
  cached_write: number | null;
  reasoning: number | null;
  image_input: number | null;
  audio_input: number | null;
  audio_output: number | null;
  embedding: number | null;
}

export interface BatchPricing {
  input: number | null;
  output: number | null;
}

export interface ModelPricing {
  id: string;
  provider: string;
  model_id: string;
  model_name: string;
  pricing: Pricing;
  batch_pricing: BatchPricing | null;
  context_length: number | null;
  max_output_tokens: number | null;
  is_open_source: boolean | null;
  capabilities: string[];
  input_modalities: string[];
  output_modalities: string[];
  last_updated: string;
}

export interface PricingUpdate {
  input?: number | null;
  output?: number | null;
  cached_input?: number | null;
}

export interface ModelUpdate {
  context_length?: number | null;
  max_output_tokens?: number | null;
  is_open_source?: boolean | null;
  pricing?: PricingUpdate;
  capabilities?: string[];
}

export interface ProviderInfo {
  name: string;
  display_name: string;
  model_count: number;
  last_updated: string | null;
}

export interface Stats {
  total_models: number;
  providers: number;
  avg_input_price: number;
  avg_output_price: number;
  last_refresh: string;
}

export type ViewMode = 'table' | 'card';

export type SortField = 'model_name' | 'input' | 'output' | 'context_length';

export interface SortConfig {
  field: SortField;
  order: 'asc' | 'desc';
}

export interface Filters {
  provider: string | null;
  capability: string | null;
  family: string | null;
  search: string;
}

export interface ModelFamily {
  name: string;
  count: number;
}
